### PR TITLE
refactor the retry logic.

### PR DESCRIPTION
**Pull request by junchaowu**
@blampe  @abhinav @breerly 
Take big method into parts

### CODE REVIEW

**Comment by junchaowu** [comment]
The github CR sucks

File: python/tchannel/tornado/peer.py [hunk 1]
```diff
@@ -551,22 +548,40 @@ def send(self, arg1, arg2, arg3,
...
+        log.debug("Got response %s", response)
+        raise gen.Return(response)
+
+    @gen.coroutine
+    def _send(self, connection, req):
+        response_future = connection.send_request(req)
+        with timeout(response_future, req.ttl):
+            response = yield response_future
+
+        raise gen.Return(response)
```
Comment (abhinav) [inline python/tchannel/tornado/peer.py]: No need for this whole thing to be a coroutine. Just do:

    return gen.with_timeout(connection.send_request(req), req.ttl)

File: python/tchannel/tornado/peer.py [hunk 2]
```diff
@@ -551,22 +548,40 @@ def send(self, arg1, arg2, arg3,
...
-                response = yield self._send(
-                    connection, request)
+                response = yield self._send(connection, request)
                 break
             except (ProtocolError, TimeoutError) as e:
+                # stop the outgoing request
                 request.set_exception(e)
-                if not request.should_retry_on_error(e):
+                if not request.should_retry_on_error(e) or (
+                        num_of_attempt == attempt_times - 1):
```
Comment (abhinav) [inline python/tchannel/tornado/peer.py]: Style nit: If the condition is becoming ungainly, move it into an aptly named variable.

    should_retry = (
        request.should_retry_on_error(e) and
        num_of_attempt <= ...
    )

    if not should_retry:
        ...

File: python/tchannel/tornado/peer.py (hunk 1 above)
Comment (junchaowu) [inline python/tchannel/tornado/peer.py]: discussed offline. Keep it.

**Comment by junchaowu** [comment]
Update the patch.

**Comment by abhinav** [comment]
:+1:

**Pull request merged**